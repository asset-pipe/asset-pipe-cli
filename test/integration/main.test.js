'use strict';

const express = require('express');
const bodyParser = require('body-parser');

let mockWriter;
const mockPlugin = jest.fn();
const mockTransform = jest.fn();

function closeServer(server) {
    return new Promise(resolve => {
        server.close(resolve);
    });
}

jest.mock('asset-pipe-js-writer', () => {
    const { Readable } = require('stream');
    mockWriter = jest.fn(() => ({
        plugin: mockPlugin,
        transform: mockTransform,
        bundle: () => {
            const items = [{}, {}, {}, {}];
            return new Readable({
                objectMode: true,
                read() {
                    if (!items.length) {
                        return this.push(null);
                    }
                    this.push(items.shift());
                },
            });
        },
    }));
    return mockWriter;
});

jest.mock('asset-pipe-css-writer', () => {
    const { Readable } = require('stream');
    const items = [{}, {}, {}, {}];
    const cssWriter = new Readable({
        objectMode: true,
        read() {
            if (!items.length) {
                return this.push(null);
            }
            this.push(items.shift());
        },
    });
    return jest.fn(() => cssWriter);
});

const Client = require('../../');

function createTestServer(handlers) {
    const server = express();
    server.use(bodyParser.json());
    for (const handler of handlers) {
        server[handler.verb](handler.path, handler.cb);
    }
    return new Promise(resolve => {
        const serve = server.listen(() => {
            resolve({
                server: serve,
                port: serve.address().port,
            });
        });
    });
}

test('uploadFeed(files, options) - js', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/js',
            cb(req, res) {
                res.send({ message: 'Success!' });
            },
        },
    ]);
    const fakeFiles = ['first.js', 'second.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const result = client.uploadFeed(fakeFiles, fakeOptions);

    await expect(result).resolves.toEqual({ message: 'Success!' });
    await closeServer(server);
});

test('uploadFeed(files, options) - js - uses transforms', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/js',
            cb: (req, res) => res.send('Success!'),
        },
    ]);
    const fakeFiles = ['first.js', 'second.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const fakeTransform = { _id: 1 };
    const fakeTransformOptions = { _id: 2 };
    client.transform(fakeTransform, fakeTransformOptions);

    await client.uploadFeed(fakeFiles, fakeOptions);

    expect(mockTransform.mock.calls[0]).toEqual([
        fakeTransform,
        fakeTransformOptions,
    ]);
    await closeServer(server);
});

test('uploadFeed(files, options) - js - uses plugins', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/js',
            cb: (req, res) => res.send({ message: 'Success!' }),
        },
    ]);
    const fakeFiles = ['first.js', 'second.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const fakePlugin = { _id: 1 };
    const fakePluginOptions = { _id: 2 };
    client.plugin(fakePlugin, fakePluginOptions);

    await client.uploadFeed(fakeFiles, fakeOptions);

    expect(mockPlugin.mock.calls[0]).toEqual([fakePlugin, fakePluginOptions]);
    await closeServer(server);
});

test('uploadFeed(files) - 200', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/js',
            cb: (req, res) => res.status(200).send({ message: 'Bad request!' }),
        },
    ]);
    const fakeFiles = ['first.js', 'second.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const result = client.uploadFeed(fakeFiles, fakeOptions);

    await expect(result).resolves.toEqual({ message: 'Bad request!' });
    await closeServer(server);
});

test('uploadFeed(files) - 400', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/js',
            cb: (req, res) => res.status(400).send({ message: 'Bad request!' }),
        },
    ]);
    const fakeFiles = ['first.js', 'second.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const result = client.uploadFeed(fakeFiles, fakeOptions);

    await expect(result).rejects.toEqual(new Error('Bad request!'));
    await closeServer(server);
});

test('uploadFeed(files) - other status codes', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/js',
            cb: (req, res) => res.status(300).send({ message: 'Bad request!' }),
        },
    ]);
    const fakeFiles = ['first.js', 'second.js'];
    const fakeOptions = {};
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const result = client.uploadFeed(fakeFiles, fakeOptions);

    await expect(result).rejects.toEqual(
        new Error(
            'Asset build server responded with unknown error. Http status 300'
        )
    );
    await closeServer(server);
});

test('uploadFeed(files) - css', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/feed/css',
            cb: (req, res) => res.status(200).send(req.body),
        },
    ]);
    const fakeFiles = ['first.css', 'second.css', 'third.css'];
    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });

    const result = client.uploadFeed(fakeFiles);

    await expect(result).resolves.toEqual([{}, {}, {}, {}]);
    await closeServer(server);
});

test('createRemoteBundle(sources) - 200', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/bundle/js',
            cb: (req, res) => res.send(req.body),
        },
    ]);
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });
    const result = client.createRemoteBundle(fakeSources, 'js');

    await expect(result).resolves.toEqual(fakeSources);
    await closeServer(server);
});

test('createRemoteBundle(sources) - css', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/bundle/css',
            cb: (req, res) => res.send(req.body),
        },
    ]);
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });
    const result = client.createRemoteBundle(fakeSources, 'css');

    await expect(result).resolves.toEqual(fakeSources);
    await closeServer(server);
});

test('createRemoteBundle(sources) - 202', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/bundle/js',
            cb: (req, res) => res.status(202).send({ message: 'success 202' }),
        },
    ]);
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });
    const result = client.createRemoteBundle(fakeSources, 'js');

    await expect(result).resolves.toEqual({ message: 'success 202' });
    await closeServer(server);
});

test('createRemoteBundle(sources) - 400', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/bundle/js',
            cb: (req, res) => res.status(400).send({ message: 'Bad request' }),
        },
    ]);
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });
    const result = client.createRemoteBundle(fakeSources, 'js');

    await expect(result).rejects.toEqual(new Error('Bad request'));
    await closeServer(server);
});

test('createRemoteBundle(sources) - other status codes', async () => {
    expect.assertions(1);
    const { server, port } = await createTestServer([
        {
            verb: 'post',
            path: '/bundle/js',
            cb: (req, res) =>
                res.status(420).send({ message: 'Its that time again' }),
        },
    ]);
    const fakeSources = ['a12das3d.json', '12da321fd.json'];

    const client = new Client({ buildServerUri: `http://127.0.0.1:${port}` });
    const result = client.createRemoteBundle(fakeSources, 'js');

    await expect(result).rejects.toEqual(
        new Error(
            'Asset build server responded with unknown error. Http status 420'
        )
    );
    await closeServer(server);
});
