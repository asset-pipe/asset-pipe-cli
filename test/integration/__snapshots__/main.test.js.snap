// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`publishAssets(tag, files, options) - 500 error 1`] = `"Error: Asset build server responded with an error. Original message: Server error. Status code: 500. URL: http://127.0.0.1publish-assets?rebundle=true"`;

exports[`publishAssets(tag, files, options) - css 1`] = `
Object {
  "data": Array [
    Object {},
    Object {},
    Object {},
    Object {},
  ],
  "tag": "podlet1",
  "type": "css",
}
`;

exports[`publishAssets(tag, files, options) - js - minify options 1`] = `
Object {
  "minify": "true",
  "rebundle": "true",
  "sourceMaps": "true",
}
`;

exports[`publishAssets(tag, files, options) - js - query params 1`] = `
Array [
  Object {
    "minify": "true",
    "rebundle": "true",
  },
  Object {
    "rebundle": "true",
  },
  Object {
    "rebundle": "true",
    "sourceMaps": "true",
  },
  Object {
    "rebundle": "true",
  },
  Object {
    "rebundle": "true",
  },
  Object {
    "rebundle": "true",
  },
]
`;

exports[`publishAssets(tag, files, options) - js - query params overrides options 1`] = `
Object {
  "rebundle": "true",
}
`;

exports[`publishAssets(tag, files, options) - js 1`] = `
Object {
  "data": Array [
    Object {},
    Object {},
    Object {},
    Object {},
  ],
  "tag": "podlet1",
  "type": "js",
}
`;

exports[`publishInstructions(tag, type, data) - 500 1`] = `[Error: Asset build server responded with an error. Original message: Server error.]`;

exports[`publishInstructions(tag, type, data) - css 1`] = `
Object {
  "data": Array [
    "a12das3d.json",
    "12da321fd.json",
  ],
  "tag": "layout",
  "type": "css",
}
`;

exports[`publishInstructions(tag, type, data) - js - options 1`] = `
Object {
  "minify": "true",
  "sourceMaps": "true",
}
`;

exports[`publishInstructions(tag, type, data) - js - query params 1`] = `
Array [
  Object {
    "minify": "true",
  },
  Object {},
  Object {
    "sourceMaps": "true",
  },
  Object {},
  Object {},
  Object {},
]
`;

exports[`publishInstructions(tag, type, data) - js - query params overrides options 1`] = `Object {}`;

exports[`publishInstructions(tag, type, data) - js 1`] = `
Object {
  "data": Array [
    "a12das3d.json",
    "12da321fd.json",
  ],
  "tag": "layout",
  "type": "js",
}
`;
